### DISCUSSION TITLE
Got stuck at "starting k3s-agent" Connecting a new node from a second devstack VM to the k3s cluster running in an instance on another devstack VM

### DISCUSSION THREAD

**Discussion by MRColorR** [k3s-io/k3s#6877] [Q&A] [answered]
Hello, i need help for an university project:

I have two Ubuntu 22.04 VMs running devstack and in the first of them I have within devstack a Debian genericCloud instance running a k3s cluster (2 nodes) . I'd like to run and add another node to this cluster, but I want to run it inside another debian instance inside the second devstack installation (i.e. on the second Ubuntu VM). But I cant find a way to make it work:
The instances from both devstack can ping each others, i can use `curl -vk https://<masterip>:6443/cacerts` to check i the instances can reach to the master node inside the other instance and they can.

if i ad another node from a new instance inside the same devstack using: `sudo curl -sfL https://get.k3s.io | K3S_URL="https://<masterInstanceFloatingIP>:6443" K3S_TOKEN="<masterNodeToken>" sh - ` it works as expected

but when i run the same command from an instance inside the second openstack it get stuck at: "starting k3s-agent" . if from this instance i `curl -vk https://<masterip>:6443/cacerts` i can see the certificate but then running `sudo k3s agent -t <token> -s https://<masterInstanceFloatingIP>:6443` i got:

> INFO[0000] Starting k3s agent v1.25.6+k3s1 (9176e03c)   
> INFO[0000] Running load balancer k3s-agent-load-balancer 127.0.0.1:6444 -> [172.16.0.168:6443 172.26.81.168:6443] 
> ERRO[0006] CA cert validation failed: Get "https://127.0.0.1:6444/cacerts": read tcp 127.0.0.1:52646->127.0.0.1:6444: read: connection reset by peer 
> INFO[0012] Waiting to retrieve agent configuration; server is not ready: failed to retrieve configuration from server: Get "https://127.0.0.1:6444/v1-k3s/config": read tcp 127.0.0.1:52776->127.0.0.1:6444: read: connection reset by peer 

you can see the steps i took to execute this project on my github repo: https://github.com/MRColorR/onehundredten/tree/master inside the steps to reproduce file.

i'll really appreciate any help you can give me, thank you

**Comment by MRColorR** [comment]
maybe i found the issue:
running angain the k3s agent -t token -s https://<master1afloatingIP>:6443 i found : 

> INFO[0005] Wrote flannel subnet file to /run/flannel/subnet.env 
> INFO[0005] Running flannel backend.                     
> I0203 14:12:02.345565    5490 vxlan_network.go:61] watching for new subnet leases
> I0203 14:12:02.423431    5490 iptables.go:270] bootstrap done
> I0203 14:12:02.477463    5490 iptables.go:270] bootstrap done
> ERRO[0006] Failed to connect to proxy. Empty dialer response  error="dial tcp 172.16.0.168:6443: connect: no route to host"
> ERRO[0006] Remotedialer proxy error                      error="dial tcp 172.16.0.168:6443: connect: no route to host"
> INFO[0011] Connecting to proxy                           url="wss://172.16.0.168:6443/v1-k3s/connect"
> ERRO[0014] Failed to connect to proxy. Empty dialer response  error="dial tcp 172.16.0.168:6443: connect: no route to host"
> ERRO[0014] Remotedialer proxy error                      error="dial tcp 172.16.0.168:6443: connect: no route to host"

the master1 floating ip is 172.26.81.168 and its internal ip is 172.16.0.168.
the node during the k3s agent starting discover the ip of the k3s cluster master node (e.g. master1a), but it use the internal ip address of the master node that is nor reachable from ouside the first openstack. it should use the floatingIP i provided in the  --server flag but then i decide to communicate to the node using the iternal ip. 

how can i force it to contact the master node to the floating ip?

**Reply by brandond** [reply]
Server nodes need a static IP address; if the "floating" IP changes periodically then this will not work. If the address is actually bound to the interface, you can try starting the server with the --node-ip flag to use this address, but if there is any NAT in place this probably will not work.

For things to work well, you should try to get all the nodes on the same network segment, with fixed addresses.

**Reply by MRColorR** [reply]
Thank you for your reply 
Yes the nodes has a floating ip address assigned dy devstack and bounded to each node's internal ip. the floating ip assigned should not change and as the floating ip are inside the same /20 network they can ping and connect each other , i tried also to get the certificate from the server node usign curl on the candidate worker node and it worked. 

I'll try with the `--node-ip` flag as i think the issue is caused by the server node advertising to the others node its internal ip instead of the floating one as from its point of view the internal IP is the one it has on eth0 interface. I thougth that passing to the additional workers node during the k3s agent join the floating ip of the master as `--server` parameter would suffice.

**Reply by MRColorR** [reply]
ok i managed to connect the nodes from my other devstack vm:
1) on devstack1 VM there's a debian genericCloud instance in wich i ran the command to setup the master node of the cluster: 
`curl -sfL https://get.k3s.io | INSTALL_K3S_EXEC="--node-external-ip <masterFloatingIP>" sh -`
2) on devstack1 VM there's a debian genericCloud instance in wich i ran the command to setup the worker node of the cluster using:
`curl -sfL https://get.k3s.io | K3S_URL=https://<masterFloatingIP>:6443 K3S_TOKEN="token" INSTALL_K3S_EXEC="--node-external-ip <workerFloatingIP>" sh -`
the node join the cluster and its state became ready.  di i used the `--node-ip` flag the node started but in the agent node i saw 

> E0204 11:23:26.095844    1173 kubelet_node_status.go:595] "Failed to set some node status fields" err="failed to validate nodeIP: node IP: \"172.26.86.169\" not found in the host's network interfaces" node="worker2b"

So i tried to use the `--node-external-ip`  insteda of  `--node-ip` and no errors appeared.
Do you think this is a possible solution?

Also i'm reading this https://docs.k3s.io/installation/network-options#distributed-hybrid-or-multicloud-cluster and i'm thinking if i need to use these flags in my case.

Then i tried to deploy my blueprint and it worked but still something is not working as expected:
a) ~~metric server get stats only from the master node and other nodes running in instances in the same openstack1 VM, others nodes remains "unknown" state if i use `sudo kubectl top nodes` , so maybe some additional ports need to be opened or some flag need to be used to point to the correct floating ip?~~ Fixed adding the --flannel-external-ip flag to the server node.

b) load balancing not working as i can access my website only if connect to the floatingIP of the node running its replicas.

**Comment by MRColorR** [comment, marked as answer]
i can now connect the nodes from both devstacks instances using this settings

1. On Devstack VM1 and VM2 enable proxy_arp and ip_forwarding
2. use two different floating ip ranges on devstack VM1 and VM2, both sould be inside the same network adress space
3. start a debian instance in devstack VM1 and assign it a floating ip, do the same on devstack VM2
4. both instances are now reachable byech other and also by any device connected to the same network
5. on instance inside devstack1 start K3s using: `curl -sfL https://get.k3s.io | INSTALL_K3S_EXEC="--node-external-ip <masterFloatingIP> --flannel-external-ip" sh -`
6. on instance inside devstack2 join the cluster using: `curl -sfL https://get.k3s.io | K3S_URL=https://<masterFloatingIP>:6443 K3S_TOKEN="token" INSTALL_K3S_EXEC="--node-external-ip <workerFloatingIP>" sh -`
7. check from the master node that bot nodes are visible and in ready state, also the metric server receive info from each node (thanks to the flannel-external-ip flag ) 

**so all seems good** ~~but there's still an issue:  after deploying my blueprint that expose a website based on nginx image using a nodeport , i can access from any device on the network but only if i use the floating ip address of the node running the pod , normally any node floating ip should work. why is this happening ?~~ you can see my project topology on my github: https://github.com/MRColorR/onehundredten/blob/master/ProjectTopology.txt
